File: android/src/main/java/com/reactnativecompressor/Audio/AudioHelper.kt
package com.reactnativecompressor.Audio

import android.media.MediaExtractor
import android.media.MediaFormat
import com.facebook.react.bridge.ReadableMap
import com.reactnativecompressor.Utils.Utils
import java.io.File
import java.io.IOException


class AudioHelper {

  var quality: String? = "medium"
  var bitrate: Int = -1
  var samplerate: Int = -1
  var channels: Int = -1
  var progressDivider: Int? = 0

  companion object {
    fun fromMap(map: ReadableMap): AudioHelper {
      val options = AudioHelper()
      val iterator = map.keySetIterator()
      while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        when (key) {
          "quality" -> options.quality = map.getString(key)
          "bitrate" -> {
            val bitrate = map.getInt(key)
            options.bitrate = if (bitrate > 320000 || bitrate < 64000) 64000 else bitrate
          }
          "samplerate" -> options.samplerate = map.getInt(key)
          "channels" -> options.channels = map.getInt(key)
        }
      }
      return options
    }


    fun getAudioBitrate(path: String): Int {
      val file = File(path)
      val fileSize = file.length() * 8 // size in bits

      val mex = MediaExtractor()
      try {
        mex.setDataSource(path)
      } catch (e: IOException) {
        e.printStackTrace()
      }

      val mf = mex.getTrackFormat(0)
      val durationUs = mf.getLong(MediaFormat.KEY_DURATION)
      val durationSec = durationUs / 1_000_000.0 // convert duration to seconds

      return (fileSize / durationSec).toInt()/1000 // bitrate in bits per second
    }
    fun getDestinationBitrateByQuality(path: String, quality: String): Int {
      val originalBitrate = getAudioBitrate(path)
      var destinationBitrate = originalBitrate
      Utils.addLog("source bitrate: $originalBitrate")

      when (quality.lowercase()) {
        "low" -> destinationBitrate = maxOf(64, (originalBitrate * 0.3).toInt())
        "medium" -> destinationBitrate = (originalBitrate * 0.5).toInt()
        "high" -> destinationBitrate = minOf(320, (originalBitrate * 0.7).toInt())
        else -> Utils.addLog("Invalid quality level. Please enter 'low', 'medium', or 'high'.")
      }

      return destinationBitrate
    }

  }
}

File: ios/Audio/AudioHelper.swift
//
//  AudioHelper.swift
//  react-native-compressor
//
//  Created by Numan on 12/11/2023.
//

import AVFoundation

class AudioHelper {
    
    static func getAudioBitrate(path: String) -> Int {
        let audioURL = URL(fileURLWithPath: path)
        let avAsset = AVURLAsset(url: audioURL)
        let keys: Set<URLResourceKey> = [.totalFileSizeKey, .fileSizeKey]
        let resourceValues = try? audioURL.resourceValues(forKeys: keys)
        let fileSize = resourceValues?.fileSize ?? resourceValues?.totalFileSize

        // Calculate bitrate in kbps
        if let fileSize = fileSize, avAsset.duration.seconds > 0 {
            return Int(Double(fileSize) * 8 / avAsset.duration.seconds)
        }
        return 0
    }
    
    static func getDestinationBitrateByQuality(path: String, quality: String) -> Int {
            let originalBitrate = getAudioBitrate(path: path)
            var destinationBitrate = originalBitrate
            print("source bitrate: \(originalBitrate)")

            // Calculate the percentage of the original bitrate relative to the range 64000 to 320000
            let percentage = Double(originalBitrate - 64000) / Double(320000 - 64000)

            switch quality.lowercased() {
            case "low":
                destinationBitrate = max(64000, Int(Double(originalBitrate) * 0.3))
            case "medium":
                // Set destination bitrate to 60% of the original bitrate
                destinationBitrate = Int(Double(originalBitrate) * 0.5)
            case "high":
                // Set destination bitrate to 80% of the original bitrate
                destinationBitrate = min(320000,Int(Double(originalBitrate) * 0.7))
            default:
                print("Invalid quality level. Please enter 'low', 'medium', or 'high'.")
            }
            
            return destinationBitrate
        }
}
